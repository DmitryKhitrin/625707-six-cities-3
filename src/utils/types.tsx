import {Action} from "redux";
import {ThunkDispatch} from "redux-thunk";
import {AxiosInstance} from "axios";

export type ThunkAction<R, S, E, A extends Action> = (
  dispatch: ThunkDispatch<S, E, A>,
  getState: () => S,
  api: AxiosInstance
) => R;
