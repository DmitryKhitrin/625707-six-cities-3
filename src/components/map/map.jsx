import React, {PureComponent} from "react";
import {PropTypes} from "prop-types";
import leaflet from "leaflet";

export class Map extends PureComponent {
  constructor(props) {
    super(props);
  }

  _initCities() {
    const {placeCardsList, city} = this.props;

    const icon = leaflet.icon({
      iconUrl: `img/pin.svg`,
      iconSize: [30, 30]
    });

    const zoom = 12;
    const map = leaflet.map(`map`, {
      center: city,
      zoom,
      zoomControl: false,
      marker: true
    });
    map.setView(city, zoom);
    leaflet
      .tileLayer(
          `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
          {
            attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
          }
      )
      .addTo(map);

    placeCardsList.map((card) => {
      leaflet.marker(card.coords, {icon, title: card.cardName}).addTo(map);
    });
  }

  componentDidMount() {
    this._initCities();
  }

  render() {
    return <div id="map" style={{height: 1000}} />;
  }
}

Map.propTypes = {
  city: PropTypes.arrayOf(PropTypes.number).isRequired,
  placeCardsList: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        priceValue: PropTypes.number,
        placeCardImage: PropTypes.string,
        cardName: PropTypes.string,
        starsRating: PropTypes.string,
        roomType: PropTypes.string,
        isPremium: PropTypes.bool,
        coords: PropTypes.arrayOf(PropTypes.number).isRequired
      })
  )
};

