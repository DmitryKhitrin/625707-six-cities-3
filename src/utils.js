import {sortTypes} from "./sortTypes.js";

const parseRating = (rate) => (String(Math.abs(rate * 100 / 5)) + `%`);

const parseUser = ({avatar_url: hostPhoto, name: hostName, id, is_prop: isProp}) => ({
  hostPhoto,
  hostName,
  id,
  isProp,
});

export const parseOffer = ({
  preview_image: previewImage,
  images,
  title,
  is_favorite: isFavorite,
  is_premium: isPremium,
  rating,
  type,
  bedrooms,
  max_adults: maxAdults,
  price,
  goods,
  host,
  location,
  description,
  id,
  city
}) => {
  return {
    previewImage,
    images,
    title,
    isFavorite,
    isPremium,
    rating: parseRating(rating),
    type,
    bedrooms,
    maxAdults,
    price,
    goods,
    host: parseUser(host),
    location: [location.latitude, location.longitude],
    description,
    id: String(id),
    city: city.name
  };
};

export const parseCities = (offers) => {
  const citiesList = new Set();
  return offers.map((offer) => {
    const {city} = offer;
    const {name, location} = city;
    return {
      name,
      location: [parseFloat(location.latitude), parseFloat(location.longitude)],
    };
  }
  ).filter((city) => {
    if (!citiesList.has(city.name)) {
      citiesList.add(city.name);
      return true;
    }
    return false;
  });
};

export const parseFavorites = (favorites) => {
  if (favorites.length === 0) {
    return {};
  }
  return favorites.reduce((acc, offer) => {
    const parsedOffer = parseOffer(offer);
    const {city} = parsedOffer;
    if (acc[city]) {
      acc[city] = [...acc[city], parsedOffer];
    } else {
      acc[city] = [parsedOffer];
    }
    return acc;
  }, {});
};

const rateToNumber = (rate) => (Number(rate.replace(`%`, ``)));

export const sortOffers = (type, offersList) => {
  switch (type) {
    case sortTypes.LOW_TO_HIGHT:
      return [...offersList].sort((a, b) => a.price - b.price);
    case sortTypes.HIGHT_TO_LOW:
      return [...offersList].sort((a, b) => b.price - a.price);
    case sortTypes.TOP_RATED:
      return [...offersList].sort(
          (a, b) => rateToNumber(b.rating) - rateToNumber(a.rating)
      );
    default:
      return [...offersList];
  }
};

export const parseComment = ({
  comment,
  date,
  id,
  rating,
  user,
}) => ({
  comment,
  date,
  id: String(id),
  rating: parseRating(rating),
  user: parseUser(user)
});

