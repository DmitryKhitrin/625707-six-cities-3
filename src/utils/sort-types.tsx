export const SORT_TYPES = {
  HIGHT_TO_LOW: `Price: high to low`,
  LOW_TO_HIGHT: `Price: low to high`,
  POPULAR: `Popular`,
  TOP_RATED: `Top rated first`,
};
