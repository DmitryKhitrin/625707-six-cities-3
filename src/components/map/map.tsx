import React, {PureComponent} from "react";
import leaflet from "leaflet";
import {ParsedOfferCard, ParsedCity} from "../../utils/utils";


type Props = {
  placeCardsList: Pick<ParsedOfferCard, |`id` | `title` | `location`>[];
  city: ParsedCity;
  height: number;
  activeCard: string;
}

const SETTINGS = {
  zoom: 13,
};

export class Map extends PureComponent<Props> {

  _map: leaflet.Map = {} as leaflet.Map;
  _zoom = 13;
  _markersLayer: leaflet.LayerGroup<any> = {} as leaflet.LayerGroup<any>;
  _activeCard = ``;

  _getIcon() {
    return leaflet.icon({
      iconUrl: `img/pin.svg`,
      iconSize: [30, 30],
    });
  }

  _getActiveIcon() {
    return leaflet.icon({
      iconUrl: `img/pin-active.svg`,
      iconSize: [30, 30],
    });
  }

  _initCities() {
    const {city} = this.props;
    const {location = [52.38333, 4.9]} = city;
    this._map = leaflet.map(`map`, {
      center: location,
      zoom: SETTINGS.zoom,
      zoomControl: false,
    });
    this._map.setView(location, this._zoom);
    leaflet
             .tileLayer(
                 `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
                 {
                   attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
                 }
             )
             .addTo(this._map);

    this._addMarkers();
  }

  _addMarkers() {
    const {placeCardsList, activeCard} = this.props;
    this._activeCard = activeCard;
    this._markersLayer = leaflet.layerGroup().addTo(this._map);

    placeCardsList.forEach((card) => {
      const {location, title, id} = card;
      if (!location) {
        return;
      }
      const icon =
        this._activeCard === id ? this._getActiveIcon() : this._getIcon();
      leaflet
        .marker(location, {icon, title})
        .addTo(this._markersLayer);
    }

    );
  }

  componentDidMount() {
    this._initCities();
  }

  componentDidUpdate() {
    const {name, location} = this.props.city;
    if (name && location.length) {
      this._markersLayer.clearLayers();
      this._map.setView(location, this._zoom);
      this._addMarkers();
    }

  }

  render() {
    return <div id="map" style={{height: this.props.height}} />;
  }
}
