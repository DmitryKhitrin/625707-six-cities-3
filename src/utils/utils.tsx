/* eslint-disable camelcase */
import {SortTypes} from "./sort-types";

export type Location = {
  latitude: number;
  longitude: number;
}

export type City = {
name: string;
location: Location;
}

export type ParsedCity = {
  name: string;
  location: [number, number]
}

const parseCity = ({name, location}: City): ParsedCity => ({
  name,
  location: [location.latitude, location.longitude],
});

const parseRating = (rate: number): string => (String(Math.abs(rate * 100 / 5)) + `%`);

export type User = {
  avatar_url: string;
  name: string;
  id: number;
  is_pro: boolean;
}

export type ParsedUser = {
  personPhoto: string;
  personName: string;
  id: number;
  isPro: boolean;
}

const parseUser = ({avatar_url: personPhoto, name: personName, id, is_pro: isPro}: User): ParsedUser => ({
  personPhoto,
  personName,
  id,
  isPro,
});

export type OfferCard = {
  preview_image: string;
  images: string[];
  title: string;
  is_favorite: boolean;
  is_premium: boolean;
  rating: number;
  type: string;
  bedrooms: number;
  max_adults: number;
  price: number;
  goods: string[];
  host: User;
  location: Location;
  description: string;
  id: number;
  city: City;
}

export type ParsedOfferCard = {
  previewImage: string;
  images: string[];
  title: string;
  isFavorite: boolean;
  isPremium: boolean;
  rating: string;
  type: string;
  bedrooms: number;
  maxAdults: number;
  price: number;
  goods: string[];
  host: ParsedUser;
  location: [number, number]
  description: string;
  id: string;
  city: ParsedCity;
}

export const parseOffer = ({
  preview_image: previewImage,
  images,
  title,
  is_favorite: isFavorite,
  is_premium: isPremium,
  rating,
  type,
  bedrooms,
  max_adults: maxAdults,
  price,
  goods,
  host,
  location,
  description,
  id,
  city
}: OfferCard): ParsedOfferCard => {
  return {
    previewImage,
    images,
    title,
    isFavorite,
    isPremium,
    rating: parseRating(rating),
    type,
    bedrooms,
    maxAdults,
    price,
    goods,
    host: parseUser(host),
    location: [location.latitude, location.longitude],
    description,
    id: String(id),
    city: parseCity(city)
  };
};

export const parseCities = (offers: OfferCard[]): ParsedCity[] => {
  const citiesList = new Set();
  return offers.map((offer) => {
    const {city} = offer;
    return parseCity(city);
  }
  ).filter((city) => {
    if (!citiesList.has(city.name)) {
      citiesList.add(city.name);
      return true;
    }
    return false;
  });
};


export type FavoriteMap = Record<string, ParsedOfferCard[]>;

export const parseFavorites = (favorites: OfferCard[]): FavoriteMap => {
  if (favorites.length === 0) {
    return {};
  }
  return favorites.reduce((acc, offer) => {
    const parsedOffer = parseOffer(offer);
    const {city: {
      name,
    }} = parsedOffer;
    if (acc[name]) {
      acc[name] = [...acc[name], parsedOffer];
    } else {
      acc[name] = [parsedOffer];
    }
    return acc;
  }, {} as FavoriteMap);
};

const rateToNumber = (rate: string): number => (Number(rate.replace(`%`, ``)));

export const sortOffers = (type: string, offersList: ParsedOfferCard[]): ParsedOfferCard[] => {
  switch (type) {
    case SortTypes.LOW_TO_HIGHT:
      return [...offersList].sort((a, b) => a.price - b.price);
    case SortTypes.HIGHT_TO_LOW:
      return [...offersList].sort((a, b) => b.price - a.price);
    case SortTypes.TOP_RATED:
      return [...offersList].sort(
          (a, b) => rateToNumber(b.rating) - rateToNumber(a.rating)
      );
    default:
      return [...offersList];
  }
};

export type Comment = {
  comment: string;
  date: string;
  id: number;
  rating: number;
  user: User;
}

export type ParsedComment = {
  comment: string;
  date: string;
  id: string;
  rating: string;
  user: ParsedUser;
}

export const parseComment = ({
  comment,
  date,
  id,
  rating,
  user,
}: Comment): ParsedComment => ({
  comment,
  date,
  id: String(id),
  rating: parseRating(rating),
  user: parseUser(user)
});

