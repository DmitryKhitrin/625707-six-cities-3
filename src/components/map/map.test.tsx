import React from "react";
import renderer from "react-test-renderer";
import {Map} from "./map";
import {ParsedOfferCard} from "../../utils/utils";

const mocData = [
  {
    id: `1`,
    price: 1,
    previewImage: `img/apartment-01.jpg`,
    rating: `25%`,
    type: `Apartment`,
    isPremium: true,
    title: `Just Flat.`,
    isFavorite: false,
    setFavorite: () => {}
  },
  {
    id: `2`,
    price: 20,
    previewImage: `img/room.jpg`,
    rating: `50%`,
    type: `Just Flat`,
    isPremium: false,
    title: `Just plane text.`,
    isFavorite: false,
    setFavorite: () => {}
  },
  {
    id: `3`,
    price: 100,
    previewImage: `img/apartment-02.jpg`,
    rating: `100%`,
    type: `Private room`,
    isPremium: true,
    title: `Best place in the World really.Beautiful luxurious apartment at great location`,
    isFavorite: false,
  }
] as ParsedOfferCard[];


it(`<Map /> should render map.`, () => {
  Map.prototype.componentDidMount = jest.fn();
  const tree = renderer
    .create(<Map city={{name: `Moscow`, location: [52.38333, 4.9]}} placeCardsList={mocData} height={1000} activeCard={``}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
