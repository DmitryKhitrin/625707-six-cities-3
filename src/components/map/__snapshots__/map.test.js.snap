// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> should render map. 1`] = `
<div
  id="map"
  style={
    Object {
      "height": 1000,
    }
  }
/>
`;
